blockchain/state_manager/src/lib.rs

```diff
@@ -122,6 +138,83 @@ where
+            if let Some(cid_pair) = self.cache.read().await.get(&tipset.key()) {
+                return Ok(cid_pair.clone());
+            }
+
+            if tipset.len() == 0 {
+                // NB: This is here because the process that executes blocks requires that the
+                // block miner reference a valid miner in the state tree. Unless we create some
+                // magical genesis miner, this won't work properly, so we short circuit here
+                // This avoids the question of 'who gets paid the genesis block reward'
+                let message_receipts = tipset
+                    .blocks()
+                    .first()
+                    .ok_or_else(|| Error::Other("Could not get message receipts".to_string()))?;
+                let cid_pair = (
+                    tipset.parent_state().clone(),
+                    message_receipts.message_receipts().clone(),
+                );
+                self.cache
+                    .write()
+                    .await
+                    .insert(tipset.key().clone(), cid_pair.clone());
+                return Ok(cid_pair);
+            }
+
+            let block_headers = tipset.blocks();
+            // generic constants are not implemented yet this is a lowcost method for now
+            let cid_pair = self.compute_tipset_state(&block_headers)?;
+            self.cache
+                .write()
+                .await
+                .insert(tipset.key().clone(), cid_pair.clone());
+            Ok(cid_pair)
+        })
+    }
+
+    pub fn compute_tipset_state<'a>(
+        &'a self,
+        blocks: &[BlockHeader],
+    ) -> Result<(Cid, Cid), Box<dyn StdError>> {
+        span!("compute_tipset_state", {
+            let check_for_duplicates = |s: &BlockHeader| {
+                blocks
+                    .iter()
+                    .filter(|val| val.miner_address() == s.miner_address())
+                    .take(2)
+                    .count()
+            };
+            if blocks.iter().any(|s| check_for_duplicates(s) > 1) {
+                // Duplicate Miner found
+                return Err(Box::new(Error::Other(
+                    "Could not get message receipts".to_string(),
+                )));
+            }
+
+            let chain_store = ChainStore::new(self.bs.clone());
+            let blocks = blocks
+                .iter()
+                .map::<Result<Block, Box<dyn StdError>>, _>(|s: &BlockHeader| {
+                    let (bls_messages, secp_messages) = chain_store.messages(&s)?;
+                    Ok(Block {
+                        header: s.clone(),
+                        bls_messages,
+                        secp_messages,
+                    })
+                })
+                .collect::<Result<Vec<Block>, _>>()?;
+            // convert tipset to fulltipset
+            let full_tipset = FullTipset::new(blocks)?;
+            let chain_rand = ChainRand::new(full_tipset.to_tipset().key().to_owned());
```

Comment:
This is cloning all tipset headers (indication from `to_tipset`) when you really only need to generate Cids and should be able to do that through references. Could you do that in this PR or is there a reason not to?

Reply:
That makes sense, I think it was unexperience with the codebase and it made sense form a Type perspective.

I used the block_header but had to add a cloned because TipsetKeys needs owned Cids